File: built-in plugins dsp/mb-delay-analog.cpp
/**
 * MB Analog Delay
 * Category : effect
 * Type     : delay
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : BBD-style analog delay
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_DELAY_ANALOG_H
#define MB_DELAY_ANALOG_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbDelayAnalog : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-delay-analog";
    static constexpr const char* PLUGIN_NAME    = "MB Analog Delay";
    static constexpr const char* PLUGIN_TYPE    = "delay";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float time = 200f;  // range [20, 800]
    float feedback = 0.5f;  // range [0, 0.9]
    float color = 0.5f;  // range [0, 1]
    float mix = 0.35f;  // range [0, 1]
    };

    MbDelayAnalog() = default;
    ~MbDelayAnalog() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.time = std::clamp(params.time, 20f, 800f);
        params.feedback = std::clamp(params.feedback, 0f, 0.9f);
        params.color = std::clamp(params.color, 0f, 1f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Analog Delay
        return input;
    }
};

#endif // MB_DELAY_ANALOG_H

File: built-in plugins dsp/mb-delay-analog.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbDelayAnalogPlugin: PluginDefinition = { id: 'mb-delay-analog', slug: 'mb-delay-analog', name: 'MB Analog Delay', category: 'effect', type: 'delay', version: '1.0.0', description: 'BBD-style analog delay', author: 'Max Booster', parameters: [{ id: 'time', name: 'Time', type: 'float', defaultValue: 200, minValue: 20, maxValue: 800, automatable: true }, { id: 'feedback', name: 'Feedback', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 0.9, automatable: true }, { id: 'color', name: 'Color', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 0.35, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { time: 200, feedback: 0.5, color: 0.5, mix: 0.35 } };

export default MbDelayAnalogPlugin;
